### PR TITLE
Add 9-slice background support to button formspec elements

### PR DESCRIPTION
**Pull request by Df458**
This adds the `bgimg_middle` style property for the `button`/`image_button`/`item_image_button` formspec elements. When set, this property makes the button's background image render as in 9-slice mode and defines the middle rectangle in the same format as the `background9` formspec element.

The result looks like this:
Screenshot-2020-01-07-18:07:11

This feature is very useful for styling formspecs, because it allows modders to skin all of their buttons with a single texture that will render consistently regardless of button size or aspect ratio.

This PR results in a small amount of duplicated code with the existing 9-slice code (specifically, the parsing and rendering code). I'm not sure how much of this we want to merge, and how. One option is to re-use the `guiBackgroundImage` class for rendering button background images, but I'm a little worried that this may complicate future changes by mixing a 'non-content' element into the button's children. We may be able to put the parsing code elsewhere though, I'm open to suggestions for where to put this code.

## How To Test
To test this PR, create a formspec using the new `bgimg_middle` style property on various button elements. Try testing with multiple types of buttons, with different sizes and 9-slice configurations (the rectangle can be defined by 1, 2, or 4 spacing values depending on how the middle rect is set up). Try combining this with various configurations of `bgimg`, `bgimg_hovered`, and `bgimg_pressed`.

To aid with this, I've created a test mod covering many of these cases: 
[test_9slice.zip](https://github.com/minetest/minetest/files/4035237/test_9slice.zip) This mod adds the `/9slice` command, which displays a formspec with various skinned buttons.

### CODE REVIEW

**Comment by rubenwardy** [comment]
> This mod adds the /9slice command, which displays a formspec with various skinned buttons.

This should be merged into `/formspec`

**Comment by SmallJoker** [comment]
I believe a better name would be `bgimg9` - not for aesthetic reasons but for extensibility. This way it would be easy to add "pressed" and "hover" 9-slice images without inventing other weird style names.

**Comment by Df458** [comment]
To address your comment, @SmallJoker (I can't see it here for some reason but I got the email notification):
I don't expect that we'll be adding variants for the hovered/pressed state, because one of my next goals PR-wise is to implement state-based styling (likely either just in time for 5.2.0, or for 5.3.0). When that happens the current hovered/pressed properties will be rendered obsolete and this property will be changeable per-state anyway.

**Comment by SmallJoker** [comment]
@Df458 Actually I removed the comment right after posting it due to the following doc entry:
> Makes the bgimg texture**s** render in 9-sliced mode and defines the middle rect.

Thus, I assumed this covers all states and expects all textures to be in the same format either 9-slice or regular). Yet I don't know how this "state-based styling" will be added to the API, but if it makes styling easier and more user-friendly I won't hold you back :)

**Comment by Df458** [comment]
> Thus, I assumed this covers all states and expects all textures to be in the same format either 9-slice or regular).

This is correct, if you try out the test mod you will see that all buttons are using state feedback already :)

**Comment by Df458** [comment]
@rubenwardy I'm not sure why it took almost a week and a half, but I've finally realized that your earlier comment was a request and added an example button. This PR should now be all set to review when you have the time to do so.

File: src/gui/StyleSpec.h [hunk 1]
```diff
@@ -117,6 +120,30 @@ class StyleSpec
 		return color;
 	}
 
+	irr::core::rect<s32> getRect(Property prop, irr::core::rect<s32> def) const
+	{
+		const auto &val = properties[prop];
+		if (val.empty()) {
```
Comment (rubenwardy) [inline src/gui/StyleSpec.h]: brace removal possible

Comment (Df458) [reply src/gui/StyleSpec.h]: Done

**Pull request merged**